src/shared/treeview/resource.ts

```diff
@@ -0,0 +1,54 @@
+/*!
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import { assign } from '../utilities/collectionUtils'
+import { TreeItemContent, TreeNode } from './resourceTreeDataProvider'
+
+export interface Resource {
+    /**
+     * The identifier associated with the resource.
+     *
+     * This should be considered _globally_ unique, trancending conventional references. Consumers of
+     * the interface can and most likely will treat this identifier as canonical.
+     */
+    readonly id: string
+}
+
+export interface ResourceProvider<T extends Resource = Resource> {
+    listResources(): Promise<T[]> | T[]
+    readonly onDidChange?: vscode.Event<void>
+}
+
+export class ResourceTreeNode<T extends Resource> implements TreeNode<T> {
```

Comment:
This isn't used in this PR, though I included it anyway to demonstrate how boilerplate involved with `TreeNode` could be removed. You could obviously keep packing in more logic here for error handling, pagination, etc.